main/java/mjc/beans/AbstractPurchase.java

```diff
@@ -0,0 +1,76 @@
+package main.java.mjc.beans;
+
+import main.java.mjc.exceptions.CsvLineException;
+
+public abstract class AbstractPurchase implements Comparable<AbstractPurchase>, Cloneable {
+    private final Product product;
+    private final int amount;
+
+    public AbstractPurchase(String[] fields) throws Exception {
```

Comment:
```suggestion
    public AbstractPurchase(String[] fields) {
```
to throw `Exception` from anywhere is antipattern.   
to throw **checked** exceptions from a ctor is a bad idea.
recommendation from J. Bloch is to throw some suitable **unchecked** exception from jdk, for example, `IllegalArgumentException` or your own subclass instance from it.